recipe/recipe.go

```diff
@@ -36,3 +42,335 @@ type instance struct {
 type file struct {
 	Description string `bson:"description,omitempty" json:"description,omitempty"`
 }
+
+//validateInstance - checks if fields of OutputInstances are not empty for ValidateAddRecipe and ValidateAddInstance
+func (instance *Instance) validateInstance() (missingFields []string, invalidFields []string) {
+
+	if instance.DatasetID == "" {
+		missingFields = append(missingFields, "dataset-id")
+	}
+
+	if instance.Editions != nil && len(instance.Editions) > 0 {
+		for j, edition := range instance.Editions {
+			if edition == "" {
+				missingFields = append(missingFields, "editions["+strconv.Itoa(j)+"]")
+			}
+		}
+	} else {
+		missingFields = append(missingFields, "editions")
+	}
+
+	if instance.Title == "" {
+		missingFields = append(missingFields, "title")
+	}
+
+	if instance.CodeLists != nil && len(instance.CodeLists) > 0 {
+		for i, codelist := range instance.CodeLists {
+			codelistMissingFields, codelistInvalidFields := codelist.validateCodelists()
+
+			if len(codelistMissingFields) > 0 {
+				for _, mField := range codelistMissingFields {
+					codelistMissingFields[i] = "codelists[" + strconv.Itoa(i) + "]." + mField
+				}
+			}
+			if len(codelistInvalidFields) > 0 {
+				for _, iField := range codelistInvalidFields {
+					codelistInvalidFields[i] = "codelists[" + strconv.Itoa(i) + "]." + iField
+				}
+			}
+
+			missingFields = append(missingFields, codelistMissingFields...)
+			invalidFields = append(invalidFields, codelistInvalidFields...)
+		}
+	} else {
+		missingFields = append(missingFields, "codelists")
+	}
+
+	return missingFields, invalidFields
+}
+
+//validateCodelists - checks if fields of CodeList are not empty for ValidateAddRecipe, ValidateAddInstance, ValidateAddCodelist
+func (codelist *CodeList) validateCodelists() (missingFields []string, invalidFields []string) {
+	hrefURL := "http://localhost:22400/code-lists/"
+
+	if codelist.ID == "" {
+		missingFields = append(missingFields, "id")
+	}
+
+	if codelist.HRef == "" {
+		missingFields = append(missingFields, "href")
+	} else {
+		if codelist.HRef != hrefURL+codelist.ID {
+			invalidFields = append(invalidFields, "href should be in format (URL/id)")
+		}
+	}
+
+	if codelist.Name == "" {
+		missingFields = append(missingFields, "name")
+	}
+
+	if codelist.IsHierarchy == nil {
+		missingFields = append(missingFields, "isHierarchy")
+	}
+
+	return missingFields, invalidFields
+}
+
+//ValidateAddRecipe - checks if all the fields of the recipe are non-empty
+func (recipe *Response) ValidateAddRecipe() error {
+	var missingFields []string
+	var invalidFields []string
+	validFormats := map[string]bool{
```

Comment:
consider moving this map to be a const or its own centrally defined type. This isnt a map we'd want the app to ever be able to change, and having a central source of truth for this would be useful

Reply:
We can't make a map to be a const. But instead, validFormats is now a global var in which it can be accessed by the functions in the package

Done in commit "optimising code by adding sub-functions and global var"